---
title: "Untitled"
author: "Your Name Here"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
#setwd("/projectnb/dispevol/E_Schlatter/kernelSGS/analysis")
setwd("C:/Users/eschlatter/Dropbox/DispersalSGS/analysis")
#treefile = "../output/FSTs_300k/ts_8238198510834684392_t300000"
treefile = "../output/500k/ts_8349801080707846925_t500"

library(dplyr)
library(stringr)
library(readr)
library(tidyr)

library(vcfR)
library(adegenet)
library(dartR)
library(snpStats)
#library(HardyWeinberg)
```

Issues with loading dartR, because of dependency on SNPRelate. Grabbed this from here: https://github.com/green-striped-gecko/dartR/issues/60
```{r, eval=FALSE}
# library(devtools)
# install.packages("BiocManager")
# BiocManager::install(c("SNPRelate", "qvalue"))
# install_github("green-striped-gecko/dartR")
# library(dartR)

# Also need snpStats (from Bioconductor):
# library(BiocManager)
# BiocManager::install("snpStats")
```

## Read in Individuals data from tskit
This is very messy and ugly, but it gets there. The id's are preserved, and the conversion to numeric didn't mess anything up.
```{r}
#############################################################################
# read in Individuals data from tskit
#############################################################################
inds <- read_csv(paste0(treefile,".csv"),col_names=F,show_col_types=FALSE)
inds <- as.data.frame(t(inds))
inds <- separate_wider_delim(inds,cols=V1,delim=',',names=c('flags','x','y','zero','parents','array','dtype','pedID','pedp1','pedp2','age','subpop','sex','flags2')) %>%
  dplyr::select(x,y,pedID,pedp1,pedp2,age,sex)
inds <- mutate(inds,x=str_trim(str_sub(x,start=18)),pedID=str_trim(str_sub(pedID,-8,-1)),
               pedp1=str_trim(str_sub(pedp1,-8,-1)),pedp2=str_trim(str_sub(pedp2,-8,-1)),
               age=str_trim(str_sub(age,-1,-1)),sex=str_trim(str_sub(sex,-1,-1)))
inds <- mutate(inds,id=0:(nrow(inds)-1),x=as.numeric(x),y=as.numeric(y),age=as.numeric(age))
```

## Read in sample sites data
```{r}
#############################################################################
# read in empirical sample sites data
#############################################################################
sample_sites <- read_csv('../data/sgs_sites_SLiM',show_col_types=FALSE)
sample_inds <- data.frame(site=numeric(),id=numeric(),dist=numeric(),x=numeric(),y=numeric()) # to store the sampled individuals

# then run Fst, using a list of the lists of Node ID's as the sample_sets
for(site in 1:nrow(sample_sites)){
  if(sample_sites$n_SNP[site]>0){
    site_x <- sample_sites$x[site]
    site_y <- sample_sites$y[site]
    inds$dist <- sqrt((inds$x-site_x)^2+(inds$y-site_y)^2)
    #  inds[[paste0('site',site,'dist')]] <- sqrt((inds$x-site_x)^2+(inds$y-site_y)^2)
    inds_for_site <- slice_min(inds,order_by=dist,n=sample_sites$n_SNP[site])
    temp.df <- data.frame(site=sample_sites$Site[site],
                        id=inds_for_site$id,
                        dist=inds_for_site$dist,
                        x = inds_for_site$x,
                        y = inds_for_site$y)
    sample_inds <- rbind(sample_inds,temp.df)  
  }
}

which(duplicated(sample_inds$id)) # check for individuals included in more than one sample; there are none



#write.csv(sample_inds,file=paste0(treefile,"_sample_inds.csv")) # save this for later reference of which individuals were sampled
```

## Can we read in the VCF file when it's big?

```{r}
library(pegas)
```

```{r}
treefile <- "../output/ts_8238198510834684392_t160000"
sample_inds <- read_csv(paste0(treefile,"_sample_inds.csv"))

f1 <- paste0(treefile,".vcf")
simVCF <- read.vcfR(f1)


VCFheader(f1)
vcf_ids <- VCFlabels(f1)
#vcf_loci <- VCFloci(f1)

simVCF <- read.vcfR(f1,nrows=1,cols=c(10), verbose=TRUE)

sample_inds <- mutate(sample_inds,gt_name=paste0("tsk_",id))
col_keep <- colnames(simVCF@gt) %in% sample_inds$gt_name
```



## Read in VCF file from tskit

```{r}
#############################################################################
# read in VCF file from tskit
#############################################################################

simVCF <- read.vcfR(paste0(treefile,".vcf"),verbose=FALSE)
#simVCF <- read.vcfR(paste0(treefile,".vcf"),cols=c(1:8,9:12),verbose=FALSE)

# vcf contains 3 elements: @meta, @fix, @gt.
# @meta is metadata
# @fix and @gt are tabular: one row for each variant (SNP), some columns of data about the variants (in @fix), and one column for each individual in the sample (in @gt).

# pull out just the columns of @gt that match sample_inds$id, and store them in vcf_sub@gt
sample_inds <- mutate(sample_inds,gt_name=paste0("tsk_",id))
col_keep <- colnames(simVCF@gt) %in% sample_inds$gt_name
col_keep[1] <- TRUE #keep first column of @gt (called FORMAT)
simVCF@gt <- simVCF@gt[,col_keep]

# create a factor of the population of each individual in vcf_sub
ind_sites <- as.factor(arrange(sample_inds,id)$site)
```

## Convert to genind and add population

```{r}
#############################################################################
# Convert to genind, add population info, convert to genlight
#############################################################################

simGI <- vcfR2genind(simVCF) #vcfR
simGI@pop <- ind_sites
simGL <- gi2gl(simGI,verbose=0) #dartR
```




## Map -- that's cool!
```{r}
load('../data/mmp_crop.RData') # use this to convert x and y into lat and lon
sample_inds <- mutate(sample_inds,lon = min_lon+x*(max_lon-min_lon)/dist_x, lat = min_lat+y*(max_lat-min_lat)/dist_y)

coords <- arrange(sample_inds,id)[,c('lat','lon')]
simGL@other$latlon <- coords
gl.map.interactive(simGL, ind.circle.cols='black',ind.circle.cex=1)
```


## Filter

```{r}
#############################################################################
# Filter: LD and minor alleles
#############################################################################

# LD:
simGL@pop <- as.factor(rep('pop1',length(simGL@ind.names))) #remove site info temporarily
ld_report <- gl.report.ld.map(simGL,maf=0.02)
simGL <- gl.filter.ld(simGL,ld_report, threshold=0.8) # R^2>0.8 is what was used in the paper. But they used PLINK, which does a sliding window thing...(?)
simGL@pop <- ind_sites #restore site info

# HWE
# I think I'm going to skip this. It's normally meant to remove loci under selection, and there is no selection in this simulation. Probably what it's actually doing is removing loci with signs of population subdivision, which I specifically don't want to do.
# simGL <- gl.filter.hwe(simGL,subset='each')

# rare alleles
simGL <- gl.filter.maf(simGL,threshold=0.02) # also removes monomorphic loci
# after filtering for rare alleles: 1118 genotypes, 219 binary SNPs
```


## Calculate pairwise FST

```{r}
#############################################################################
# Calculate pairwise fsts
#############################################################################

fsts <- gl.fst.pop(simGL,nboots=1)  #dartR
save(fsts,file=paste0(treefile,"_FST.RData"))
```


## adegenet: DAPC?
I'm kind of blindly following instructions; I don't really know what I'm doing. So I don't trust any of it.
```{r}

simGI <- gl2gi(simGL)
save(simGI,file=paste0(treefile,'_filter_genind.RData'))

lgpop <- substr(simGI@pop,start=1,stop=1)

grp <- find.clusters(simGI,max.n.clust=50)
grp <- find.clusters(simGI,n.pca=300,n.clust=3)

table(lgpop,grp$grp)
table.value(table(lgpop,grp$grp), col.lab=paste("inf", 1:3), row.lab=c('B','G','L','T'))


dapc1 <- dapc(simGI,lgpop,n.pc=22,n.da=3)
scatter(dapc1,cstar=0, scree.da=FALSE,cellipse=0)
```


